### PR TITLE
xdg-autostart-service: handle gnome autostart phase better on other d…

### PR DESCRIPTION
**Pull request by davidedmundson**
Autostart files which contain the line gnome-autostart-phase are currently completely skipped by systemd. This is because these are handled internally by gnome startup through other means.

The problem is a number of desktop files that need to run on kde too have this flag set. Ideally they should just create systemd user units, but we're not at this point universally yet.

This patch changes the logic so if the flag is set, we set NotShowIn-gnome, which in turn would just not load decided at runtime.

As an optimisation if we would get conflicting OnlyShowIn lines we still skip the file completely.

### CODE REVIEW

**Comment by davidedmundson** [comment]
@benzea 

CC: https://github.com/systemd/systemd/issues/18791

File: src/xdg-autostart-generator/xdg-autostart-service.c [hunk 1]
```diff
@@ -562,9 +571,23 @@ int xdg_autostart_service_generate_unit(
         }
 
         if (service->gnome_autostart_phase) {
-                /* There is no explicit value for the "Application" phase. */
-                log_debug("%s: not generating unit, startup phases are not supported.", service->path);
-                return 0;
+                /* There is no explicit value for the "Application" phase.
+                 *
+                 * On gnome secondary startup mechanism handles desktop files with startup phases set.
+                 * We want to mark these as "NotShowIn=gnome"
```
Comment (benzea) [inline src/xdg-autostart-generator/xdg-autostart-service.c]: nit: NotShowIn=GNOME (uppercase)

**Review by benzea — COMMENTED** [review]
looks good to me.

File: src/xdg-autostart-generator/xdg-autostart-service.c [hunk 2]
```diff
@@ -562,9 +571,23 @@ int xdg_autostart_service_generate_unit(
...
+                 * On gnome secondary startup mechanism handles desktop files with startup phases set.
+                 * We want to mark these as "NotShowIn=GNOME"
+                 *
+                 * If that means no-one will load them, we can get skip it entirely
+                 */
+                if (strv_contains(only_show_in, "GNOME")) {
+                        strv_remove(only_show_in, "GNOME");
+
+                        if (strv_isempty(only_show_in)) {
+                                log_debug("%s: Gnome startup phases are handled separately on GNOME. Skipping", service->path);
```
Comment (YHNdnzj) [inline src/xdg-autostart-generator/xdg-autostart-service.c]: The first `Gnome` is unnecessary? (Or at least use proper cases?)

File: src/xdg-autostart-generator/xdg-autostart-service.c [hunk 3]
```diff
@@ -562,9 +571,23 @@ int xdg_autostart_service_generate_unit(
...
+                                return 0;
+                        }
+                }
+                log_debug("%s: Gnome startup phases are handled seprately on GNOME, marking as NotShowIn=GNOME", service->path);
```
Comment (YHNdnzj) [inline src/xdg-autostart-generator/xdg-autostart-service.c]: Same

File: src/xdg-autostart-generator/xdg-autostart-service.c [hunk 4]
```diff
@@ -562,9 +571,23 @@ int xdg_autostart_service_generate_unit(
...
+                strv_extend(&not_show_in, "GNOME");
```
Comment (poettering) [inline src/xdg-autostart-generator/xdg-autostart-service.c]: oom check...

**Comment by YHNdnzj** [comment]
ping?

**Comment by keszybz** [comment]
I'll rebase the patch.

**Comment by keszybz** [comment]
Updated with the requested changes and some trivial cleanups.

**Review by YHNdnzj — APPROVED** [review]

**Comment by benzea** [comment]
Thanks everyone!

**Pull request merged**